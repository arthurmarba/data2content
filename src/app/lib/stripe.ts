import Stripe from "stripe";
import { assertBillingEnv } from "@/app/lib/boot-sanity";

assertBillingEnv();

// Garante `string` no tipo (e no runtime)
const secret: string = (() => {
  const v = process.env.STRIPE_SECRET_KEY;
  if (!v) {
    throw new Error("STRIPE_SECRET_KEY ausente nas variáveis de ambiente.");
  }
  return v;
})();

// Versão basil (pin do stripe@18.4.0)
const DEFAULT_API_VERSION: Stripe.LatestApiVersion = "2025-07-30.basil";

// Usa env se fornecida; caso contrário, trava na basil
const apiVersion: Stripe.LatestApiVersion =
  (process.env.STRIPE_API_VERSION as Stripe.LatestApiVersion | undefined) ??
  DEFAULT_API_VERSION;

// Em produção, exige a mesma versão para evitar drift
if (process.env.NODE_ENV === "production" && apiVersion !== DEFAULT_API_VERSION) {
  throw new Error(
    `STRIPE_API_VERSION divergente (${apiVersion}). Use ${DEFAULT_API_VERSION}.`
  );
}

let _stripe: Stripe | null = null;

/**
 * Retorna sempre a MESMA instância do cliente Stripe.
 * Evita múltiplas conexões/configs divergentes.
 */
export function getStripe(): Stripe {
  if (!_stripe) {
    _stripe = new Stripe(secret, { apiVersion });
  }
  return _stripe;
}

// atalho conveniente
export const stripe = getStripe();
