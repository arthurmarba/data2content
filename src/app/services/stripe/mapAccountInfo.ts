// src/app/services/stripe/mapAccountInfo.ts
import type Stripe from 'stripe';

export interface StripeUiStatus {
  payoutsEnabled: boolean;
  needsOnboarding: boolean;
  disabledReasonKey?: string;
  defaultCurrency?: string;
  accountCountry?: string;
  isUnderReview?: boolean;
}

export function mapStripeAccountInfo(account: Stripe.Account): StripeUiStatus {
  const payoutsEnabled = !!account.payouts_enabled;
  const defaultCurrency = account.default_currency
    ? String(account.default_currency).toUpperCase()
    : undefined;
  const accountCountry = account.country
    ? String(account.country).toUpperCase()
    : undefined;
  const disabledReasonKey =
    (account.requirements as any)?.disabled_reason ||
    (account as any).disabled_reason ||
    undefined;
  const needsOnboarding =
    !payoutsEnabled || (account.requirements?.currently_due?.length ?? 0) > 0;
  const isUnderReview =
    disabledReasonKey === 'under_review' ||
    (disabledReasonKey?.startsWith('rejected.') ?? false);

  return {
    payoutsEnabled,
    needsOnboarding,
    disabledReasonKey,
    defaultCurrency,
    accountCountry,
    isUnderReview,
  };
}
