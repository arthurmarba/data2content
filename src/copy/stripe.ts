export const STRIPE_STATUS = {
  verified: 'Verificado',
  action_needed: 'Ação necessária',
  under_review: 'Em revisão',
} as const;

export const STRIPE_DISABLED_REASON: Record<string, {title: string, body: string, cta?: 'onboarding'|'contact'|'wait'}> = {
  'requirements.past_due': {
    title: 'Informações pendentes',
    body: 'Faltam documentos obrigatórios no Stripe. Envie-os para liberar seus saques.',
    cta: 'onboarding',
  },
  'requirements.pending_verification': {
    title: 'Verificação em andamento',
    body: 'O Stripe está verificando seus documentos. Isso pode levar algum tempo.',
    cta: 'wait',
  },
  'rejected.fraud': {
    title: 'Conta rejeitada',
    body: 'Sua conta foi rejeitada por suspeita de fraude. Entre em contato com o suporte do Stripe.',
    cta: 'contact',
  },
  'rejected.terms_of_service': {
    title: 'Conta rejeitada',
    body: 'Houve um problema com os Termos do Stripe. Fale com o suporte do Stripe para detalhes.',
    cta: 'contact',
  },
  'under_review': {
    title: 'Conta em revisão',
    body: 'Sua conta está passando por uma revisão interna do Stripe.',
    cta: 'wait',
  },
  // fallback genérico
  default: {
    title: 'Ação necessária no Stripe',
    body: 'Há uma pendência na sua conta Stripe. Revise seu cadastro para liberar os saques.',
    cta: 'onboarding',
  },
};

export const CURRENCY_HELP = {
  mismatch_banner: (balCur: string, dstCur: string) =>
    `Você tem saldo em ${balCur}. Sua conta Stripe recebe em ${dstCur}.`,
  mismatch_reason: (balCur: string, dstCur: string) =>
    `Sua conta Stripe recebe em ${dstCur}; este saldo está em ${balCur}.`,
  options_title: (balCur: string) => `Como sacar ${balCur}?`,
  options: [
    'Usar outra conta Stripe que receba na mesma moeda.',
    'Ajustar sua conta junto ao Stripe para aceitar a moeda (quando o país permitir).',
    'Solicitar ajuda ao suporte para alternativas.',
  ],
} as const;

