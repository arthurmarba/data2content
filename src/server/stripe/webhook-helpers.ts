import { User } from "@/server/db/models/User";
import { AffiliateInvoiceIndex, AffiliateSubscriptionIndex, AffiliateRefundProgress } from "@/server/db/models/AffiliateIndexes";
import type Stripe from "stripe";

export async function findUserByCustomerId(customerId: string) {
  return User.findOne({ stripeCustomerId: customerId });
}

export async function markEventIfNew(user: any, eventId: string): Promise<boolean> {
  if (user.lastProcessedEventId === eventId) return false;
  user.lastProcessedEventId = eventId;
  await user.save();
  return true;
}

export async function ensureInvoiceIdempotent(invoiceId: string, affiliateUserId: string) {
  try {
    await AffiliateInvoiceIndex.create({ invoiceId, affiliateUserId });
    return true;
  } catch {
    return false;
  }
}

export async function ensureSubscriptionFirstTime(subscriptionId: string, affiliateUserId: string) {
  try {
    await AffiliateSubscriptionIndex.create({ subscriptionId, affiliateUserId });
    return true;
  } catch {
    return false;
  }
}

export function cents(n: number) { return Math.round(n ?? 0); }

export function calcCommissionCents(invoice: Stripe.Invoice) {
  const rate = Number(process.env.COMMISSION_RATE_BPS ?? "1000") / 10000; // default 10%
  const paid = Number(invoice.amount_paid ?? 0);
  return cents(paid * rate);
}

export function addDays(date: Date, days: number) {
  const d = new Date(date); d.setDate(d.getDate() + days); return d;
}
