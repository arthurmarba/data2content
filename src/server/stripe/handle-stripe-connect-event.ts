import type Stripe from 'stripe';
import { connectToDatabase } from '@/app/lib/mongoose';
import { User } from '@/server/db/models/User';
import Redemption from '@/app/models/Redemption';
import { adjustBalance } from '@/server/affiliate/balance';

export async function handleStripeConnectEvent(event: Stripe.Event) {
  await connectToDatabase();

  switch (event.type) {
    case 'account.updated': {
      const account = event.data.object as Stripe.Account;
      const accountId = account.id;
      if (!accountId) return;

      const user = await User.findOne({ 'paymentInfo.stripeAccountId': accountId });
      if (!user) return;

      user.paymentInfo ||= {};

      const payoutsEnabled = !!account.payouts_enabled;
      const chargesEnabled = !!account.charges_enabled;
      const defaultCurrency = account.default_currency?.toUpperCase();
      const disabledReason =
        (account.requirements as any)?.disabled_reason ||
        (account as any).disabled_reason ||
        null;
      const needsOnboarding = !!(
        account.requirements?.currently_due?.length ||
        account.requirements?.past_due?.length ||
        account.requirements?.eventually_due?.length
      );
      const status = disabledReason
        ? 'disabled'
        : payoutsEnabled && chargesEnabled && !!account.details_submitted
        ? 'verified'
        : 'pending';

      user.paymentInfo.stripeAccountStatus = status;
      user.paymentInfo.stripeAccountDefaultCurrency = defaultCurrency;
      user.paymentInfo.stripeAccountPayoutsEnabled = payoutsEnabled;
      user.paymentInfo.stripeAccountChargesEnabled = chargesEnabled;
      user.paymentInfo.stripeAccountDisabledReason = disabledReason;
      user.paymentInfo.stripeAccountCapabilities = {
        card_payments: account.capabilities?.card_payments,
        transfers: account.capabilities?.transfers,
      } as any;
      user.paymentInfo.stripeAccountNeedsOnboarding = needsOnboarding;
      user.paymentInfo.stripeAccountCountry = account.country ?? null;

      await user.save();
      break;
    }
    // ATUALIZAÇÃO APLICADA AQUI
    case 'transfer.reversed': {
      const transfer = event.data.object as Stripe.Transfer;
      const accountId = event.account as string | undefined;
      if (!accountId) return;

      const user = await User.findOne({ 'paymentInfo.stripeAccountId': accountId });
      if (!user) return;

      const redemption = await Redemption.findOne({ transferId: transfer.id });
      if (redemption && redemption.status === 'paid') {
        await adjustBalance(user, transfer.currency, transfer.amount);
        redemption.status = 'rejected';
        redemption.reasonCode = 'transfer_reversed'; // Este código já estava correto
        await redemption.save();
        await user.save();
      }
      break;
    }
    case 'payout.failed': {
      const payout = event.data.object as Stripe.Payout;
      const accountId = event.account as string | undefined;
      // CORREÇÃO APLICADA AQUI para resolver o erro de tipo
      const transferId =
        typeof (payout as any).source_transaction === 'string'
          ? (payout as any).source_transaction
          : null;
      if (!accountId || !transferId) return;

      const user = await User.findOne({ 'paymentInfo.stripeAccountId': accountId });
      if (!user) return;

      const redemption = await Redemption.findOne({ transferId });
      if (redemption && redemption.status === 'paid') {
        await adjustBalance(user, payout.currency, payout.amount);
        redemption.status = 'rejected';
        redemption.reasonCode = 'payout_rejected';
        await redemption.save();
        await user.save();
      }
      break;
    }
    default:
      break;
  }
}
