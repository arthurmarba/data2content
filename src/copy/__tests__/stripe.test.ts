import { STRIPE_STATUS, STRIPE_DISABLED_REASON, CURRENCY_HELP } from '../stripe';

describe('stripe copy', () => {
  it('matches snapshot', () => {
    expect({ STRIPE_STATUS, STRIPE_DISABLED_REASON, CURRENCY_HELP }).toMatchSnapshot();
  });
});
