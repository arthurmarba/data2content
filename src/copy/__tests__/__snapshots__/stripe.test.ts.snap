// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stripe copy matches snapshot 1`] = `
{
  "CURRENCY_HELP": {
    "mismatch_banner": [Function],
    "mismatch_reason": [Function],
    "options": [
      "Usar outra conta Stripe que receba na mesma moeda.",
      "Ajustar sua conta junto ao Stripe para aceitar a moeda (quando o país permitir).",
      "Solicitar ajuda ao suporte para alternativas.",
    ],
    "options_title": [Function],
  },
  "STRIPE_DISABLED_REASON": {
    "default": {
      "body": "Há uma pendência na sua conta Stripe. Revise seu cadastro para liberar os saques.",
      "cta": "onboarding",
      "title": "Ação necessária no Stripe",
    },
    "rejected.fraud": {
      "body": "Sua conta foi rejeitada por suspeita de fraude. Entre em contato com o suporte do Stripe.",
      "cta": "contact",
      "title": "Conta rejeitada",
    },
    "rejected.terms_of_service": {
      "body": "Houve um problema com os Termos do Stripe. Fale com o suporte do Stripe para detalhes.",
      "cta": "contact",
      "title": "Conta rejeitada",
    },
    "requirements.past_due": {
      "body": "Faltam documentos obrigatórios no Stripe. Envie-os para liberar seus saques.",
      "cta": "onboarding",
      "title": "Informações pendentes",
    },
    "requirements.pending_verification": {
      "body": "O Stripe está verificando seus documentos. Isso pode levar algum tempo.",
      "cta": "wait",
      "title": "Verificação em andamento",
    },
    "under_review": {
      "body": "Sua conta está passando por uma revisão interna do Stripe.",
      "cta": "wait",
      "title": "Conta em revisão",
    },
  },
  "STRIPE_STATUS": {
    "action_needed": "Ação necessária",
    "under_review": "Em revisão",
    "verified": "Verificado",
  },
}
`;
